File: hw1/project1cs360s2020.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <utility>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <sstream>
#include <iomanip>
#include <fstream>

using namespace std;

using State = vector< pair<int, int> >;
using Action = pair<int, int>;
using Drone = pair<int, int>;
using Coord = pair<int, int>;

// #define DEBUG

struct CoordHash {
    size_t operator()(const Coord &coord) const {
        return hash<int>()(coord.first) ^ hash<int>()(coord.second);
    }
};

struct StateHash {
    size_t operator()(const State &state) const {
        size_t x = state.size();
        for(Coord coord : state) {
            x ^= CoordHash()(coord) + 0x9e3779b9 + (x << 6) + (x >> 2);
        }
        return x;
    }
};

class Problem {
public:
    virtual vector<Action> actions(State state) = 0;
    virtual State step(State state, Action action) = 0;
    virtual int solve() = 0;
    virtual int getNumPackagesCollected(State state) = 0;
    virtual int getNumEligiblePackages(State state) = 0;
    virtual string getStateString(State state) = 0;
    virtual const int getN() = 0;
    virtual const int getD() = 0;
    virtual const int getXY(Coord coord) = 0;
};

class Solver {
public:
    Solver(Problem *problem) : problem(problem) {}

    virtual int solve() = 0;

protected:
    Problem* problem;
};

class DfsSolver : public Solver {
public:
    DfsSolver(Problem* problem) : Solver(problem), numPlaced(0), 
                numPackagesCollected(0), numExpansions(0) {
        n = problem -> getN();
        d = problem -> getD();
        packages = vector< vector<int> >(n, vector<int>(n, 0));
        rowHistory.push_back(0);
        colHistory.push_back(0);
        diag1History.push_back(0);
        diag2History.push_back(0);
        placed = vector< vector<bool> >(n, vector<bool>(n, false));
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                packages[i][j] = problem -> getXY(make_pair(i,j));
                if(packages[i][j] > 0) {
                    packagesList.push_back(make_pair(i,j));
                }
            }
        }
        optSoFar = 0;
    }
    
    int solve() {
        #ifdef DEBUG
        cout << problem->getStateString(history) << endl;
        #endif

        numExpansions++;

        if(numPlaced == d) {
            return numPackagesCollected;
        }

        int opt = 0;

        int upperBound = numPackagesCollected + getOverestimation();
        if(upperBound <= optSoFar) {
            return 0;
        }

        int startingRow = (history.empty()) ? 0 : (history[history.size() - 1].first + 1);
        int endRow = n - (d - numPlaced);
        for(int i = startingRow; i <= endRow; i++) {
            for(int j = 0; j < n; j++) {
                if(!getCovered(i,j)) {
                    // stepping
                    step(make_pair(i,j));
                    opt = max(opt, solve());
                    unstep(make_pair(i,j));
                    if(opt > optSoFar) {
                        optSoFar = opt;
                    }
                    if(opt == upperBound) {
                        return opt;
                    }
                }
            }
        }

        return opt;
    }

    int getNumExpansions() {
        return numExpansions;
    }

private:
    vector< vector<int> > packages;
    vector<int> rowHistory;
    vector<int> colHistory;
    vector<int> diag1History;
    vector<int> diag2History;
    vector< vector<bool> > placed;
    vector< Action > history;
    vector< Coord > packagesList;
    int n;
    int d;
    int numPlaced;
    int numPackagesCollected;
    int numExpansions;
    int optSoFar;

    void step(Action action) {
        // create new cover ints
        int rowCoverage = rowHistory[rowHistory.size() - 1];
        rowCoverage |= 1 << action.first;
        rowHistory.push_back(rowCoverage);

        int colCoverage = colHistory[colHistory.size() - 1];
        colCoverage |= 1 << action.second;
        colHistory.push_back(colCoverage);

        int diag1Coverage = diag1History[diag1History.size() - 1];
        diag1Coverage |= 1 << (action.first - action.second + n - 1);
        diag1History.push_back(diag1Coverage);

        int diag2Coverage = diag2History[diag2History.size() - 1];
        diag2Coverage |= 1 << (action.first + action.second);
        diag2History.push_back(diag2Coverage);
        
        // update drone matrix
        placed[action.first][action.second] = true;

        // update history
        history.push_back(action);

        // update numPlaced
        numPlaced++;

        // update numPackagesCollected
        numPackagesCollected += packages[action.first][action.second];
    }

    void unstep(Action action) {
        rowHistory.pop_back();
        colHistory.pop_back();
        diag1History.pop_back();
        diag2History.pop_back();

        placed[action.first][action.second] = false;

        history.pop_back();

        numPlaced--;

        numPackagesCollected -= packages[action.first][action.second];
    }

    bool getCovered(int i, int j) {
        bool covered = false;
        covered |= (rowHistory[rowHistory.size() - 1] >> i) & 1;
        covered |= (colHistory[colHistory.size() - 1] >> j) & 1;
        covered |= (diag1History[diag1History.size() - 1] >> (i - j + n - 1)) & 1;
        covered |= (diag2History[diag2History.size() - 1] >> (i + j)) & 1;
        return covered;
    }

    int getOverestimation() { // O(num of packages)
        int res = 0;
        for(Coord coord : packagesList) {
            int i = coord.first;
            int j = coord.second;
            if(!getCovered(i, j)) {
                res += packages[i][j];
            }
        }
        return res;
    }
};

class AstarSolver : public Solver {
public:
    AstarSolver(Problem* problem) : Solver(problem) {}

    int solve() {
        State initState;
        auto comp = [this](State s1, State s2) {
            int g1 = problem->getNumPackagesCollected(s1);
            int g2 = problem->getNumPackagesCollected(s2);
            int f1 = g1 + heuristic(s1);
            int f2 = g2 + heuristic(s2);
            if(f1 != f2) {
                return f1 < f2;
            }
            return g1 < g2;
        };
        priority_queue<State, vector<State>, decltype(comp)> frontier(comp);
        frontier.push(initState);
        unordered_set<State, StateHash> frontierSet;
        unordered_set<State, StateHash> explored;
        while(true) {
            if(frontier.empty()) {
                // failure
                return 0;
            }
            State state = frontier.top();
            frontier.pop();
            frontierSet.erase(state);

            #ifdef DEBUG
            cout << problem->getStateString(state) << endl;
            #endif

            if(problem->getD() == state.size()) {
                return problem->getNumPackagesCollected(state);
            }
            explored.insert(state);
            for(Action action : problem->actions(state)) {
                State nextState = problem->step(state, action);

                if(explored.find(nextState) == explored.end()
                    && frontierSet.find(nextState) == frontierSet.end()) {
                    frontier.push(nextState);
                    frontierSet.insert(nextState);
                }
            }
        }
    }

private:
    int heuristic(State state) {
        return problem->getNumEligiblePackages(state);
    }
};

class DroneProblem : public Problem {
public:
    DroneProblem(int n, int d, vector<Coord> coords, string algorithm)
            : n(n), d(d), algorithm(algorithm) {
        for(Coord coord : coords) {
            auto it = packages.find(coord);
            if(it == packages.end()) {
                packages.insert(make_pair(coord, 1));
            }
            else {
                it->second++;
            }
        }
    }

    vector<Action> actions(State state) {
        vector<Action> eligibleActions;
        int startingRow = (state.empty()) ? 0 : (state[state.size() - 1].first + 1);
        int endRow = n - (d - state.size());
        vector< vector<bool> > covered = getCovered(state);
        for(int i = startingRow; i <= endRow; i++) {
            for(int j = 0; j < n; j++) {
                Action action = make_pair(i, j);
                if(!covered[action.first][action.second]) {
                    eligibleActions.push_back(action);
                }
            }
        }
        
        return eligibleActions;
    }

    State step(State state, Action action) {
        State nextState(state);
        nextState.push_back(action);
        return nextState;
    }

    int solve() {
        if(algorithm == "dfs") {
            return DfsSolver(this).solve();
        }
        else if(algorithm == "astar") {
            return AstarSolver(this).solve();
        }
        
        throw invalid_argument("algorithm must be dfs or astar.");
    }

    int getNumPackagesCollected(State state) {
        int numCollected = 0;
        for(Drone drone : state) {
            numCollected += getXY(drone);
        }
        return numCollected;
    }

    int getNumEligiblePackages(State state) {
        vector< vector<bool> > covered = getCovered(state);
        int eRow = getNumEligiblePackages(state, covered, "row");
        int eCol = getNumEligiblePackages(state, covered, "col");
        int eDiag1 = getNumEligiblePackages(state, covered, "diag1");
        int eDiag2 = getNumEligiblePackages(state, covered, "diag2");
        return min(min(eRow, eCol), min(eDiag1, eDiag2));
    }

    string getStateString(State state) {
        stringstream ss;
        ss << "Placed: " << setw(3) << state.size() << endl;
        ss << "Left: " << setw(3) << d - state.size() << endl;
        ss << "Score: " << setw(3) << getNumPackagesCollected(state) << endl;
        ss << "Eligb: " << setw(3) << getNumEligiblePackages(state) << endl;
        for(int i = 0; i < n; i++) {
            ss << separatorString() << endl;
            ss << rowString(state, i) << endl;
        }
        ss << separatorString() << endl;
        return ss.str();
    }

    const int getN() {
        return n;
    }

    const int getD() {
        return d;
    }

    const int getXY(Coord coord) {
        auto it = packages.find(coord);
        return (it != packages.end()) ? it->second : 0;
    }

private:
    int n; // board size
    int d; // number of drones
    // vector<Action> actionPriority;
    unordered_map<Coord, int, CoordHash> packages; // list of coordinates
    string algorithm;

    vector< vector<bool> > getCovered(State state) {
        /* n^2 runtime */
        vector< vector<bool> > covered = vector< vector<bool> >(n, vector<bool>(n, false));
        for(Coord coord : state) {
            int x = coord.first;
            int y = coord.second;

            for(int j = 0; j < n; j++) {
                covered[x][j] = true;
            }
            for(int i = 0; i < n; i++) {
                covered[i][y] = true;
            }
            for(int i = 0; i < n; i++) {
                int j = x + y - i;
                if(j >= 0 && j < n) {
                    covered[i][j] = true;
                }
                j = i - x + y;
                if(j >= 0 && j < n) {
                    covered[i][j] = true;
                }
            }
        }
        return covered;
    }

    /* Check whether a position in the board is covered by a drone. */
    bool checkCovered(State state, Coord coord) {
        for(Drone drone : state) {
            // same row
            if(drone.first == coord.first) return true;

            // same column
            if(drone.second == coord.second) return true;

            // same first diagonal
            if(drone.first - drone.second == coord.first - coord.second) {
                return true;
            }

            // same second diagonal
            if(drone.first + drone.second == coord.first + coord.second) {
                return true;
            }
        }

        return false;
    }

    int getNumEligiblePackages(State state, vector<vector<bool>> covered, string variation) {
        priority_queue<int> pq;

        if(variation == "row") {
            for(int i = 0; i < n; i++) {
                int maxPackageCount = 0;
                bool dronePlaced = false;
                for(int j = 0; j < n; j++) {
                    Coord coord = make_pair(i, j);
                    int coordPackageCount = getXY(coord);
                    if((!covered[i][j]) 
                        && (coordPackageCount > maxPackageCount)) {
                        maxPackageCount = coordPackageCount;
                    }
                    if(find(state.begin(), state.end(), coord) != state.end()) {
                        dronePlaced = true;
                        break;
                    }
                }
                if(!dronePlaced) {
                    pq.push(maxPackageCount);
                }
            }
        }
        else if(variation == "col") {
            for(int j = 0; j < n; j++) {
                int maxPackageCount = 0;
                bool dronePlaced = false;
                for(int i = 0; i < n; i++) {
                    Coord coord = make_pair(i, j);
                    int coordPackageCount = getXY(coord);
                    if((!covered[i][j]) 
                        && (coordPackageCount > maxPackageCount)) {
                        maxPackageCount = coordPackageCount;
                    }
                    if(find(state.begin(), state.end(), coord) != state.end()) {
                        dronePlaced = true;
                        break;
                    }
                }
                if(!dronePlaced) {
                    pq.push(maxPackageCount);
                }
            }
        }
        else if(variation == "diag1") {
            for(int dist = -(n - 1); dist <= (n - 1); dist++) {
                int maxPackageCount = 0;
                bool dronePlaced = false;
                int iMin = max(0, dist);
                int iMax = min(n - 1, dist + n - 1);
                for(int i = iMin; i <= iMax; i++) {
                    int j = i - dist;
                    Coord coord = make_pair(i, j);
                    int coordPackageCount = getXY(coord);
                    if((!covered[i][j]) 
                        && (coordPackageCount > maxPackageCount)) {
                        maxPackageCount = coordPackageCount;
                    }
                    if(find(state.begin(), state.end(), coord) != state.end()) {
                        dronePlaced = true;
                        break;
                    }
                }
                if(!dronePlaced) {
                    pq.push(maxPackageCount);
                }
            }
        }
        else if(variation == "diag2") {
            for(int sum = 0; sum <= 2 * (n - 1); sum++) {
                int maxPackageCount = 0;
                bool dronePlaced = false;
                int iMin = max(0, sum - (n - 1));
                int iMax = min(n - 1, sum);
                for(int i = iMin; i <= iMax; i++) {
                    int j = sum - i;
                    Coord coord = make_pair(i, j);
                    int coordPackageCount = getXY(coord);
                    if((!covered[i][j]) 
                        && (coordPackageCount > maxPackageCount)) {
                        maxPackageCount = coordPackageCount;
                    }
                    if(find(state.begin(), state.end(), coord) != state.end()) {
                        dronePlaced = true;
                        break;
                    }
                }
                if(!dronePlaced) {
                    pq.push(maxPackageCount);
                }
            }
        }
        else {
            throw invalid_argument("algorithm must be in [row, col, diag1, diag2].");
        }

        int res = 0;
        int numDronesLeft = d - state.size();
        for(int i = 0; i < numDronesLeft; i++) {
            res += pq.top();
            pq.pop();
        }
        return res;
    }

    string separatorString() {
        stringstream ss;
        for(int i = 0; i < n; i++) {
            ss << "+----";
        }
        ss << "+";
        return ss.str();
    }

    string rowString(State state, int i) {
        stringstream ss;
        for(int j = 0; j < n; j++) {
            Coord coord = make_pair(i, j);
            int numPackages = getXY(coord);
            bool drone = find(state.begin(), state.end(), coord) != state.end();
            string droneMark = drone ? "x" : " ";
            ss << "|" << droneMark << " " 
               << setw(2);
            if(numPackages > 0) {
                ss << numPackages;
            }
            else {
                ss << "";
            }
        }
        ss << "|";
        return ss.str();
    }
};

int main(int argc, char *argv[]) {
    ifstream fin("input.txt");

    int n, d, p;
    fin >> n >> d >> p;

    // read past endline
    string line;
    getline(fin, line);

    string algorithm;
    getline(fin, algorithm);

    vector<Coord> coords;
    for(int i = 0; i < p; i++) {
        getline(fin, line);
        int delimPosition = line.find(',');
        int x = stoi(line.substr(0, delimPosition));
        int y = stoi(line.substr(delimPosition + 1));
        coords.push_back(make_pair(x, y));
    }

    fin.close();

    DroneProblem problem(n, d, coords, algorithm);

    ofstream fout("output.txt");

    fout << problem.solve() << endl;

    fout.close();

    return 0;
}

File: hw1/project1cs360s2020.java
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

class Problem {
  private int boardSize;
  private State initState;

  public Problem(State state, String algorithm) {
    this.boardSize = state.getBoard().size();
    this.initState = state;
  }

  public State getInitState() {
    return initState;
  }

  public List<Action> actions(State state) {
    List<Action> eligibleActions = new ArrayList<>();
    int boardSize = state.getBoard().size();
    for(int i = 0; i < boardSize; i++) {
      for(int j = 0; j < boardSize; j++) {
        if(!state.getBoard().get(i).get(j).getCovered()) {
          eligibleActions.add(new Action(i, j));
        }
      }
    }
    return eligibleActions;
  }

  public State step(State state, Action action) {
    State nextState = new State(state);
    nextState.step(action);
    return nextState;
  }

  public int solve(String algorithm) {
    Solver solver;
    if(algorithm.equals("dfs")) {
      solver = new DfsSolver(this);
    }
    else if(algorithm.equals("astar")) {
      solver = new AstarSolver(this);
    }
    else {
      System.out.println("Invalid argument algorithm -- should be dfs or astar.");
      return 0;
    }
    return solver.solve();
  }
}

abstract class Solver {
  protected Problem problem;
  
  public Solver(Problem problem) {
    this.problem = problem;
  }

  protected void debug(State state) {
    System.out.println(state);
    try {
      Thread.sleep(1000);
    }
    catch(Exception e) {
      System.out.println("Input error.");
    }
  }

  abstract public int solve();
}

class DfsSolver extends Solver {
  public DfsSolver(Problem problem) {
    super(problem);
  }

  public int solve() {
    return helper(problem.getInitState(), 0);
  }

  private int helper(State state, int optSoFar) {
    // debug(state);

    if(state.getNumDronesLeft() == 0) {
      return state.getNumPackagesCovered();
    }

    // check if we will never reach optimal state from here
    if(state.getNumPackagesCovered() + state.getNumEligiblePackages() <= optSoFar) {
      return 0;
    }

    List<Action> eligibleActions = problem.actions(state);

    // sort actions so that we explore the ones covering most packages first
    eligibleActions.sort(new Comparator<Action>() {
      @Override
      public int compare(Action a1, Action a2) {
        int a1x = a1.getCoord().getX();
        int a1y = a1.getCoord().getY();
        int a2x = a2.getCoord().getX();
        int a2y = a2.getCoord().getY();

        int numPackages1 = state.getBoard().get(a1x).get(a1y).getNumPackages();
        int numPackages2 = state.getBoard().get(a2x).get(a2y).getNumPackages();
        return numPackages2 - numPackages1;
      }
    });

    List<Action> history = state.getHistory();
    Action lastAction = null;
    if(history.size() > 0) {
      lastAction = history.get(history.size() - 1);
    }

    int opt = optSoFar;
    for(Action action: eligibleActions) {
      // filter actions so that drones are placed with increasing
      // x values
      // if(lastAction != null && action.getCoord().getX() <= lastAction.getCoord().getX()) {
      //   continue;
      // }

      State nextState = problem.step(state, action);
      int resultForThisAction = helper(nextState, opt);
      if(resultForThisAction > opt) {
        opt = resultForThisAction;

        // current result reaches optimal condition in this state
        if(state.getNumPackagesCovered() + state.getNumEligiblePackages() == opt) {
          break;
        }
      }
    }
    return opt;
  }
}

class AstarSolver extends Solver {
  public AstarSolver(Problem problem) {
    super(problem);
  }

  public int solve() {
    State initState = problem.getInitState();
    Comparator<State> stateComparator = new Comparator<State>() {
      @Override
      public int compare(State s1, State s2) {
        int f1 = s1.getNumPackagesCovered() + heuristic(s1);
        int f2 = s2.getNumPackagesCovered() + heuristic(s2);
        return f2 - f1;
      }
    };
    PriorityQueue<State> frontier = new PriorityQueue<>(stateComparator);
    frontier.add(initState);
    Set<State> explored = new HashSet<State>();
    while(true) {
      if(frontier.isEmpty()) {
        // failure
        return 0;
      }
      State state = frontier.poll();
      // super.debug(state);
      if(state.getNumDronesLeft() == 0) {
        return state.getNumPackagesCovered();
      }
      explored.add(state);
      for(Action action : problem.actions(state)) {
        State nextState = problem.step(state, action);
        if((!explored.contains(nextState)) && (!frontier.contains(nextState))) {
          frontier.add(nextState);
        }
      }
    }
  }

  private int heuristic(State state) {
    return state.getNumEligiblePackages();
  }
}

class State {
  private List<List<Cell>> board;
  private List<Action> history;
  private int numDronesPlaced;
  private int numDronesLeft;

  public State(int boardSize, int droneCount, List<Coordinate> packageLocations) {
    board = new ArrayList<>();
    for (int i = 0; i < boardSize; i++) {
      board.add(new ArrayList<Cell>());
      for (int j = 0; j < boardSize; j++) {
        board.get(i).add(new Cell());
      }
    }

    history = new ArrayList<>();
    numDronesPlaced = 0;
    numDronesLeft = droneCount;

    for (Coordinate coord : packageLocations) {
      board.get(coord.getX()).get(coord.getY()).addPackage();
    }
  }

  public State(List<List<Cell>> board, List<Action> history, int numDronesPlaced, int numDronesLeft) {
    int boardSize = board.size();
    this.board = new ArrayList<>();
    for (int i = 0; i < boardSize; i++) {
      this.board.add(new ArrayList<>());
      for (int j = 0; j < boardSize; j++) {
        this.board.get(i).add(new Cell(board.get(i).get(j)));
      }
    }
    this.history = new ArrayList<>();
    for (int i = 0; i < history.size(); i++) {
      this.history.add(new Action(history.get(i)));
    }
    this.numDronesPlaced = numDronesPlaced;
    this.numDronesLeft = numDronesLeft;
  }

  public State(State other) {
    this(other.getBoard(), other.getHistory(), other.getNumDronesPlaced(), other.getNumDronesLeft());
  }

  public List<List<Cell>> getBoard() {
    return board;
  }

  public List<Action> getHistory() {
    return history;
  }

  public int getCost() {
    return history.size();
  }

  public int getNumDronesPlaced() {
    return numDronesPlaced;
  }

  public int getNumDronesLeft() {
    return numDronesLeft;
  }

  public int getNumPackagesCovered() {
    int boardSize = board.size();
    int numCovered = 0;
    for(int i = 0; i < boardSize; i++) {
      for(int j = 0; j < boardSize; j++) {
        Cell cell = board.get(i).get(j);
        if(cell.getDronePlaced()) {
          numCovered += cell.getNumPackages();
        }
      }
    }
    return numCovered;
  }

  private int getNumEligiblePackages(String variation) {
    int boardSize = board.size();

    PriorityQueue<Integer> maxHeap = new PriorityQueue<>(new Comparator<Integer>() {
      @Override
      public int compare(Integer o1, Integer o2) {
        return o2.compareTo(o1);
      }
    });

    if(variation.equals("row")) {
      // count eligible packages when considering one-per-row constraint
      for(int i = 0; i < boardSize; i++) {
        int maxPackageCount = 0;
        boolean dronePlaced = false;
        for(int j = 0; j < boardSize; j++) {
          Cell cell = board.get(i).get(j);
          if((!cell.getCovered()) && (cell.getNumPackages() > maxPackageCount)) {
            maxPackageCount = cell.getNumPackages();
          }
          if(cell.getDronePlaced()) {
            dronePlaced = true;
            break;
          }
        }
        if(!dronePlaced) {
          maxHeap.add(maxPackageCount);
        }
      }
    }
    else if(variation.equals("col")) {
      for(int j = 0; j < boardSize; j++) {
        int maxPackageCount = 0;
        boolean dronePlaced = false;
        for(int i = 0; i < boardSize; i++) {
          Cell cell = board.get(i).get(j);
          if((!cell.getCovered()) && (cell.getNumPackages() > maxPackageCount)) {
            maxPackageCount = cell.getNumPackages();
          }
          if(cell.getDronePlaced()) {
            dronePlaced = true;
            break;
          }
        }
        if(!dronePlaced) {
          maxHeap.add(maxPackageCount);
        }
      }
    }
    else if(variation.equals("diag1")) {
      for(int dist = -(boardSize - 1); dist <= boardSize - 1; dist++) {
        int maxPackageCount = 0;
        boolean dronePlaced = false;
        int iMin = Math.max(0, dist);
        int iMax = Math.min(boardSize - 1, dist + boardSize - 1);
        for(int i = iMin; i <= iMax; i++) {
          int j = i - dist;
          Cell cell = board.get(i).get(j);
          if((!cell.getCovered()) && (cell.getNumPackages() > maxPackageCount)) {
            maxPackageCount = cell.getNumPackages();
          }
          if(cell.getDronePlaced()) {
            dronePlaced = true;
            break;
          }
        }
        if(!dronePlaced) {
          maxHeap.add(maxPackageCount);
        }
      }
    }
    else if(variation.equals("diag2")) {
      for(int sum = 0; sum <= 2 * (boardSize - 1); sum++) {
        int maxPackageCount = 0;
        boolean dronePlaced = false;
        int iMin = Math.max(0, sum - (boardSize - 1));
        int iMax = Math.min(boardSize - 1, sum);
        for(int i = iMin; i <= iMax; i++) {
          int j = sum - i;
          Cell cell = board.get(i).get(j);
          if((!cell.getCovered()) && (cell.getNumPackages() > maxPackageCount)) {
            maxPackageCount = cell.getNumPackages();
          }
          if(cell.getDronePlaced()) {
            dronePlaced = true;
            break;
          }
        }
        if(!dronePlaced) {
          maxHeap.add(maxPackageCount);
        }
      }
    }

    int res = 0;
    for(int i = 0; i < numDronesLeft; i++) {
      res += maxHeap.poll();
    }
    return res;
  }

  public int getNumEligiblePackages() {
    return Math.min(
             Math.min(getNumEligiblePackages("row"),
                      getNumEligiblePackages("col")),
             Math.min(getNumEligiblePackages("diag1"),
                      getNumEligiblePackages("diag2"))
           );
  }

  public int getNumEligiblePackagesArchived() {
    int boardSize = board.size();
    ArrayList<Integer> eligiblePackages = new ArrayList<>();
    for(int i = 0; i < boardSize; i++) {
      for(int j = 0; j < boardSize; j++) {
        Cell cell = board.get(i).get(j);
        if(!cell.getCovered()) {
          eligiblePackages.add(cell.getNumPackages());
        }
      }
    }
    Collections.sort(eligiblePackages);
    int numEligible = 0;
    for(int i = 0; i < Math.min(numDronesLeft, eligiblePackages.size()); i++) {
      numEligible += eligiblePackages.get(eligiblePackages.size() - i - 1);
    }
    return numEligible;
  }

  public void step(Action action) {
    int boardSize = board.size();
    int x = action.getCoord().getX();
    int y = action.getCoord().getY();

    // add drone placement
    board.get(x).get(y).setDronePlaced();

    // add coverage in row
    for (int j = 0; j < boardSize; j++) {
      board.get(x).get(j).setCovered();
    }

    // add coverage in column
    for (int i = 0; i < boardSize; i++) {
      board.get(i).get(y).setCovered();
    }

    // add coverage in primary diagonal
    for (int i = 0; i < boardSize; i++) {
      int j = i + y - x;
      if (j >= 0 && j < boardSize) {
        board.get(i).get(j).setCovered();
      }
    }

    // add coverage in secondary diagonal
    for (int i = 0; i < boardSize; i++) {
      int j = x + y - i;
      if (j >= 0 && j < boardSize) {
        board.get(i).get(j).setCovered();
      }
    }

    history.add(action);
    numDronesLeft--;
    numDronesPlaced++;
  }

  @Override
  public int hashCode() {
    List<String> historyStr = history.stream()
                                     .map((Action ac) -> ("(" + ac.getCoord().getX() + "," + ac.getCoord().getY() + ")"))
                                     .collect(Collectors.toList());
    Collections.sort(historyStr);
    return String.join("", historyStr).hashCode();
  }

  @Override
  public boolean equals(Object obj) {
    if(this == obj) return true;
    if(obj == null) return false;
    if(getClass() != obj.getClass()) return false;
    State other = (State) obj;
    return this.hashCode() == other.hashCode();
  }

  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append(String.format("Placed: %3d\n", numDronesPlaced));
    sb.append(String.format("Left  : %3d\n", numDronesLeft));
    sb.append(String.format("Score : %3d\n", getNumPackagesCovered()));
    sb.append(String.format("Eligib: %3d\n", getNumEligiblePackages()));
    for(int i = 0; i < board.size(); i++) {
      sb.append(separatorString() + "\n");
      sb.append(rowString(i) + "\n");
    }
    sb.append(separatorString() + "\n");
    return sb.toString();
  }

  private String separatorString() {
    // returns ascii art for a row separator
    StringBuilder sb = new StringBuilder();
    for(int i = 0; i < board.size(); i++) {
      sb.append("+----");
    }
    sb.append("+");
    return sb.toString();
  }

  private String rowString(int i) {
    // returns ascii art for a row
    StringBuilder sb = new StringBuilder();
    for(int j = 0; j < board.size(); j++) {
      Cell cell = board.get(i).get(j);
      int numPackages = cell.getNumPackages();
      String droneMark = cell.getDronePlaced() ? "x" : " ";
      String coverMark = (cell.getCovered() && numPackages > 0) ? "+" : " ";
      String packageMark = (numPackages > 0) ? String.format("%2d", numPackages) : "  ";
      sb.append(String.format("|%s%s%s", droneMark, coverMark, packageMark));
    }
    sb.append("|");
    return sb.toString();
  }
}

class Cell {
  private int numPackages;
  private boolean covered;
  private boolean dronePlaced;

  public Cell() {
    numPackages = 0;
    covered = false;
    dronePlaced = false;
  }

  public Cell(int numPackages, boolean covered, boolean dronePlaced) {
    this.numPackages = numPackages;
    this.covered = covered;
    this.dronePlaced = dronePlaced;
  }

  public Cell(Cell other) {
    this(other.getNumPackages(), other.getCovered(), other.getDronePlaced());
  }

  public int getNumPackages() {
    return numPackages;
  }

  public boolean getCovered() {
    return covered;
  }

  public boolean getDronePlaced() {
    return dronePlaced;
  }

  public void addPackage() {
    numPackages += 1;
  }

  public void setCovered() {
    covered = true;
  }

  public void setDronePlaced() {
    dronePlaced = true;
  }
}

class Action {
  private Coordinate coord;

  public Action(int x, int y) {
    this.coord = new Coordinate(x, y);
  }

  public Action(Action other) {
    this.coord = new Coordinate(other.getCoord().getX(), other.getCoord().getY());
  }

  public Coordinate getCoord() {
    return coord;
  }
}

class Coordinate {
  private int x;
  private int y;

  public Coordinate(int x, int y) {
    this.x = x;
    this.y = y;
  }

  public int getX() {
    return x;
  }

  public int getY() {
    return y;
  }
}

class project1cs360s2020 {
  public static void main(String[] args) {
    Scanner input = new Scanner(System.in);

    int n = input.nextInt();
    int d = input.nextInt();
    int p = input.nextInt();

    input.nextLine(); // read past endline

    String algorithm = input.nextLine().trim();

    List<Coordinate> coords = new ArrayList<>();
    for (int i = 0; i < p; i++) {
      String coordStr = input.nextLine().trim();
      String[] xy = coordStr.split(",");
      int x = Integer.parseInt(xy[0]);
      int y = Integer.parseInt(xy[1]);
      coords.add(new Coordinate(x, y));
    }

    input.close();

    Problem problem = new Problem(new State(n, d, coords), algorithm);

    System.out.println(problem.solve(algorithm));
  }
}